File: src/de/saschahlusiak/freebloks/game/JoinDialog.java
package de.saschahlusiak.freebloks.game;

import de.saschahlusiak.freebloks.Global;
import de.saschahlusiak.freebloks.R;
import de.saschahlusiak.freebloks.controller.Spielleiter;
import android.app.Dialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.preference.PreferenceManager;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.AdapterView.OnItemSelectedListener;

public class JoinDialog extends Dialog {
	
	public interface OnJoinListener {
		public boolean OnJoin(String server, boolean players[], int game_mode, int field_size, String player_name);
	}
	
	final static int FIELD_SIZES[] = {
		13, 15, 17, 20, 23
	};
	
	CheckBox player1, player2, player3, player4;
	EditText name, server;
	Spinner field_size, game_mode;

	public JoinDialog(Context context, final OnJoinListener listener) {
		super(context);
		setContentView(R.layout.join_game_dialog);
		
		player1 = (CheckBox)findViewById(R.id.player1);
		player2 = (CheckBox)findViewById(R.id.player2);
		player3 = (CheckBox)findViewById(R.id.player3);
		player4 = (CheckBox)findViewById(R.id.player4);
		
		player1.setText(context.getResources().getStringArray(R.array.color_names)[0]);
		player2.setText(context.getResources().getStringArray(R.array.color_names)[1]);
		player3.setText(context.getResources().getStringArray(R.array.color_names)[2]);
		player4.setText(context.getResources().getStringArray(R.array.color_names)[3]);

		field_size = (Spinner) findViewById(R.id.field_size);
		field_size.setSelection(3);
		
		game_mode = (Spinner) findViewById(R.id.game_mode);
		game_mode.setSelection(Spielleiter.GAMEMODE_4_COLORS_4_PLAYERS);
		
		name = (EditText)findViewById(R.id.name);
		server = (EditText)findViewById(R.id.server);
		
		server.setText(Global.DEFAULT_SERVER_ADDRESS);
		Button okAdd = (Button)findViewById(android.R.id.button1);
		okAdd.setOnClickListener(new View.OnClickListener() {
			public void onClick(View v) {
				String s = server.getText().toString();
				String n = name.getText().toString();

				SharedPreferences prefs;
				prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
				Editor editor = prefs.edit();
				editor.putString("player_name", n);
				editor.commit();
				
				if (listener.OnJoin(s, getPlayers(), getGameMode(), getFieldSize(), n))
					dismiss();
			}
		});
		((Button)findViewById(android.R.id.button2)).setOnClickListener(new View.OnClickListener() {
			public void onClick(View v) {
				dismiss();
			}
		});
	}
	
	public void prepareJoinDialog() {
		int p = (int)(Math.random() * 4.0);
		
		player1.setChecked(p == 0);
		player2.setChecked(p == 1);
		player3.setChecked(p == 2);
		player4.setChecked(p == 3);
		
		SharedPreferences prefs;
		prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
		
		name.setText(prefs.getString("player_name", null));
		findViewById(R.id.linearLayout1).setVisibility(View.VISIBLE);
		findViewById(R.id.spinners).setVisibility(View.GONE);
		setTitle(R.string.join_game);		
	}

	public void prepareHostDialog() {
		int p = (int)(Math.random() * 4.0);
		
		player1.setChecked(p == 0);
		player2.setChecked(p == 1);
		player3.setChecked(p == 2);
		player4.setChecked(p == 3);
		
		SharedPreferences prefs;
		prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
		
		name.setText(prefs.getString("player_name", null));
		findViewById(R.id.linearLayout1).setVisibility(View.GONE);
		findViewById(R.id.spinners).setVisibility(View.VISIBLE);
		setTitle(R.string.host_game);
		
		game_mode.setOnItemSelectedListener(new OnItemSelectedListener() {

			@Override
			public void onItemSelected(AdapterView<?> adapter, View view, int position, long id) {				
				if (position == Spielleiter.GAMEMODE_2_COLORS_2_PLAYERS) {	
					player1.setEnabled(true);
					player3.setEnabled(true);
					player2.setEnabled(false);
					player4.setEnabled(false);

					if (player2.isChecked())
						player1.setChecked(true);
					if (player4.isChecked())
						player3.setChecked(true);
					player2.setChecked(false);
					player4.setChecked(false);
					field_size.setSelection(1); /* 15x15 */
				} else if (position == Spielleiter.GAMEMODE_4_COLORS_2_PLAYERS) {
					player1.setEnabled(true);
					player2.setEnabled(true);
					player3.setEnabled(false);
					player4.setEnabled(false);

					boolean e;
					e = player1.isChecked() || player3.isChecked();
					player1.setChecked(e);
					player3.setChecked(e);

					e = player2.isChecked() || player4.isChecked();
					player2.setChecked(e);
					player4.setChecked(e);
				} else {
					player1.setEnabled(true);
					player2.setEnabled(true);
					player3.setEnabled(true);
					player4.setEnabled(true);
				}
			}

			@Override
			public void onNothingSelected(AdapterView<?> arg0) {
				
			}
		});
		
		player1.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				if (game_mode.getSelectedItemPosition() == Spielleiter.GAMEMODE_4_COLORS_2_PLAYERS) {
					player3.setChecked(player1.isChecked());
				}
			}
		});
		player2.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				if (game_mode.getSelectedItemPosition() == Spielleiter.GAMEMODE_4_COLORS_2_PLAYERS) {
					player4.setChecked(player2.isChecked());
				}
			}
		});
	}

	boolean[] getPlayers() {
		boolean p[] = new boolean[4];
		p[0] = player1.isChecked();
		p[1] = player2.isChecked();
		p[2] = player3.isChecked();
		p[3] = player4.isChecked();
		if (game_mode.getSelectedItemPosition() == Spielleiter.GAMEMODE_4_COLORS_2_PLAYERS) {
			/* this would otherwise request players two times, the server would hand out 2x2 = 4 players */
			p[2] = p[3] = false;
		}
		return p;
	}
	
	public int getGameMode() {
		return (int)game_mode.getSelectedItemPosition();
	}
	
	public int getFieldSize() {
		return FIELD_SIZES[field_size.getSelectedItemPosition()];
	}

}

File: src/main/java/de/saschahlusiak/freebloks/game/JoinDialog.kt
package de.saschahlusiak.freebloks.game

import android.app.Dialog
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.*
import androidx.preference.PreferenceManager
import de.saschahlusiak.freebloks.Global
import de.saschahlusiak.freebloks.R
import de.saschahlusiak.freebloks.bluetooth.BluetoothClientToSocketThread
import de.saschahlusiak.freebloks.bluetooth.BluetoothServerThread
import de.saschahlusiak.freebloks.bluetooth.BluetoothServerThread.OnBluetoothConnectedListener
import de.saschahlusiak.freebloks.client.GameClient
import de.saschahlusiak.freebloks.model.GameConfig
import kotlinx.android.synthetic.main.join_dialog.*

@Deprecated("Convert to DialogFragment")
class JoinDialog(context: Context, private val listener: OnStartCustomGameListener) : Dialog(context, R.style.Theme_Freebloks_Light_Dialog), RadioGroup.OnCheckedChangeListener, View.OnClickListener, TextWatcher, OnBluetoothConnectedListener {
    private lateinit var name: EditText
    private lateinit var server: EditText
    private lateinit var serverType: RadioGroup
    private lateinit var hostGame: Button
    private lateinit var bluetoothList: ViewGroup
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothServer: BluetoothServerThread? = null
    private val prefs: SharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.join_dialog)
        setTitle(R.string.join_multiplayer_game)
        window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_HIDDEN)

        findViewById<View>(R.id.close_button).setOnClickListener { dismiss() }
        findViewById<View>(R.id.ok_button).setOnClickListener(this)

        hostGame = findViewById(R.id.host_game)
        hostGame.setOnClickListener {
            saveSettings()
            dismiss()
            listener.setClientName(getName())
            val config = GameConfig(null, true)
            listener.onStartClientGameWithConfig(config)
        }
        name = findViewById(R.id.name)
        server = findViewById(R.id.server_address)
        server.setText(prefs.getString("custom_server", ""))
        server.addTextChangedListener(this)

        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        bluetoothList = findViewById(R.id.bluetoothList)
        bluetoothList.visibility = View.GONE
        if (bluetoothAdapter == null) {
            // bluetooth not supported
            radioButtonBluetooth.visibility = View.GONE
        }
        serverType = findViewById(R.id.server_type)
        serverType.setOnCheckedChangeListener(this)
        radioButtonInternet.isChecked = true
    }

    override fun onStart() {
        super.onStart()

        startBluetoothServer()
    }

    override fun onStop() {
        bluetoothServer?.shutdown()
        bluetoothServer = null

        super.onStop()
    }

    private fun startBluetoothServer() {
        val adapter = bluetoothAdapter ?: return
        if (bluetoothServer == null && adapter.isEnabled) {
            bluetoothServer = BluetoothServerThread(this).also {
                it.start()
            }
        }
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        updateOkButtonEnabled()
    }

    override fun afterTextChanged(s: Editable) {}

    private fun updateOkButtonEnabled() {
        val checkedId = serverType.checkedRadioButtonId
        ok_button.isEnabled = when {
            checkedId == R.id.radioButtonWifi && customServerAddress.isEmpty() -> false
            checkedId == R.id.radioButtonBluetooth -> false
            else -> true
        }
    }

    override fun onClick(view: View) {
        listener.setClientName(getName())

        when (serverType.checkedRadioButtonId) {
            R.id.radioButtonInternet -> {
                val config = GameConfig(server = Global.DEFAULT_SERVER_ADDRESS, showLobby = true)
                listener.onStartClientGameWithConfig(config)
            }
            R.id.radioButtonWifi -> {
                if (customServerAddress.isEmpty()) return
                val config = GameConfig(server = customServerAddress, showLobby =  true)
                listener.onStartClientGameWithConfig(config)
            }
        }

        saveSettings()
        dismiss()
    }

    private fun saveSettings() {
        prefs.edit()
            .putString("player_name", getName())
            .putString("custom_server", customServerAddress)
            .apply()
    }

    fun setName(name: String?) {
        this.name.setText(name)
    }

    private fun getName(): String {
        return name.text.toString().trim { it <= ' ' }
    }

    private val customServerAddress: String
        get() = server.text.toString().trim { it <= ' ' }

    override fun onCheckedChanged(radioGroup: RadioGroup, i: Int) {
        when(i) {
            R.id.radioButtonInternet -> {
                server.visibility = View.GONE
                hostGame.visibility = View.GONE
                bluetoothList.visibility = View.GONE
            }
            R.id.radioButtonWifi -> {
                server.visibility = View.VISIBLE
                hostGame.visibility = View.VISIBLE
                bluetoothList.visibility = View.GONE

                server.requestFocus()
            }
            R.id.radioButtonBluetooth -> {
                server.visibility = View.GONE
                hostGame.visibility = View.VISIBLE
                bluetoothList.visibility = View.VISIBLE

                startBluetoothServer()
                updateBluetoothDeviceList()
            }
        }

        updateOkButtonEnabled()
    }

    private fun updateBluetoothDeviceList() {
        val deviceSelectedListener = View.OnClickListener { v ->
            listener.setClientName(getName())
            saveSettings()
            val device = v.tag as BluetoothDevice
            dismiss()
            Log.i(tag, "Device selected: " + device.name)
            listener.onConnectToBluetoothDevice(device)
        }

        bluetoothList.removeAllViews()

        val pairedDevices = bluetoothAdapter?.bondedDevices ?: emptySet()
        Log.d(tag, "Paired devices: " + pairedDevices.size)

        if ((bluetoothAdapter?.isEnabled != true) || pairedDevices.isEmpty()) {
            val v = layoutInflater.inflate(R.layout.join_bluetooth_device, bluetoothList, false)
            v.findViewById<View>(R.id.image).visibility = View.GONE
            v.findViewById<TextView>(android.R.id.text1).apply {
                setText(R.string.bluetooth_disabled)
            }
            bluetoothList.addView(v, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
            return
        }

        for (device in pairedDevices) {
            Log.d(tag, String.format("device %s [%s]", device.name, device.address))
            val v = layoutInflater.inflate(R.layout.join_bluetooth_device, bluetoothList, false).apply {
                findViewById<TextView>(android.R.id.text1).apply {
                    text = device.name
                }
                tag = device
                setOnClickListener(deviceSelectedListener)
            }
            bluetoothList.addView(v, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
        }
    }

    override fun onBluetoothClientConnected(socket: BluetoothSocket) {
        // a client has connected to us. quickly host a game and get the two together by starting the bridge
        dismiss()
        val config = GameConfig(server = null, showLobby = true)
        listener.setClientName(getName())
        listener.onStartClientGameWithConfig(config, Runnable {
            BluetoothClientToSocketThread(socket, "localhost", GameClient.DEFAULT_PORT).start()
        })
    }

    companion object {
        private val tag = JoinDialog::class.java.simpleName
    }
}